0000-interledger-protocol-4/0000-interledger-protocol-4.md

```diff
@@ -0,0 +1,159 @@
+- **Quoting is an Application Concern** - Connectors are only responsible for forwarding ILP packets and do not need to implement a [separate protocol for quoting](../0008-interledger-quoting-protocol/0008-interledger-quoting-protocol.md). Applications can use test packets to determine the exchange rate of a particular path.
+
+## Flow
+
+Before the protocol starts, the sender and receiver use a higher-level protocol to agree on the destination account and packet condition.
+
+1. Sender constructs an ILP Prepare packet with the receiver's destination account, the agreed-upon condition, and an amount and expiry of their choice. The sender may include additional data, the formatting of which is determined by the higher-level protocol.
+2. Sender sends the ILP Prepare packet to a connector over some communication channel, such as a Websocket. 
+3. Connector gets the packet and determines who the sender was based on the communication channel it was received through. The connector checks whether the sender has sufficient balance or credit to send the amount specified in the packet. If so, the connector reduces the sender's available balance by the value of the packet. If not, the connector returns an ILP Reject packet.
+4. Connector uses the destination ILP address from the packet and its local routing tables to determine which next hop to forward the packet to.
+5. Connector modifies the packet amount and expiry to apply their exchange rate and move the expiry timestamp earlier.
+6. Connector forwards the packet to the next hop, which may be another connector. All subsequent connectors go through steps 3-6 (treating the previous connector as the sender) until the packet reaches the receiver.
+7. Receiver checks the ILP Prepare packet, according to whatever is stipulated by the higher-level protocol (such as checking whether the amount received is above some minimum specified by the sender). 
+8. Receiver returns an ILP Fulfill packet, containing the preimage of the condition in the ILP Prepare packet. The receiver may include additional data, the formatting of which is determined by the higher-level protocol.
+9. Connector checks that the receiver has returned the fulfillment before the expiry in the ILP Prepare packet. If so, the connector returns the same ILP Fulfill packet to the previous connector and credits the receiver's account with the amount in the ILP Prepare packet. If the Prepare has expired, the connector returns an ILP Reject packet to the previous connector. Each connector repeats this step (treating the next connector as the receiver whom they will credit) until the Fulfill (or Reject) packet reaches the sender.
```

Comment:
I'm still unclear on if the ILP Reject packet is necessary since the previous connector will just time out before receiving the Fulfill packet. Is the reject simply a courtesy?

Reply:
I'll add a bit more of a description about that.

In short, the answer is that connectors must relay reject packets that they get. In addition, connectors can set timers for outgoing prepared and send back reject packets when they expire. I haven't thought about this too much but that's probably in the category of recommended behavior rather than mandatory functionality.

Reply:
I think right now a lot of things won't work without the ILP reject, but nobody will lose money so you could argue it's technically not necessary